src/components/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,304 @@
+      nextItems = this.filterGroupSearchResultFromItems(items, inputValue)
+    }
+    // 2. Standard flat item filtering
+    else {
+      nextItems = items.filter(item =>
+        this.filterSearchResult(item, inputValue)
+      )
+    }
+
+    return nextItems
+  }
+
+  renderItems = dropdownProps => {
+    const { items, renderItems, renderItemsAsGroups, hasGroups } = dropdownProps
+
+    const { customFilter } = this.props
+    const { inputValue } = this.state
+
+    const filterProps = {
+      items,
+      hasGroups,
+      inputValue,
+    }
+
+    let nextItems
+
+    // 1. User defined custom filtering
+    if (customFilter) {
+      nextItems = customFilter(filterProps, this.defaultFilter)
+    }
+    // 2. Default filtering
+    else {
+      nextItems = this.defaultFilter(filterProps)
+    }
+
+    // Render empty state, if empty
+    if (isItemsEmpty(nextItems)) {
+      return this.renderEmpty()
+    }
+
+    // We need to render the items slightly differently for groups.
+    // To do this, we'll rely on the function provided to use from Dropdown.
+    if (hasGroups) {
+      return renderItemsAsGroups({ items: nextItems, withIndex: !!inputValue })
+    } else {
+      return renderItems({ items: nextItems, withIndex: true })
+    }
+  }
+
+  onMenuMount = () => {
+    this.props.onMenuMount()
+    this.resetInputValue()
+  }
+
+  onMenuUnmount = () => {
+    this.props.onMenuUnmount()
+    this.resetInputValue()
+  }
+
+  getDropdownProps = () => {
+    const { className, onInputChange, noResultsLabel, ...rest } = this.props
+    const { inputValue } = this.state
+
+    const componentClassName = classNames('c-ComboBox', className)
+
+    return {
+      ...rest,
+      onMenuMount: this.onMenuMount,
+      onMenuUnmount: this.onMenuUnmount,
+      onSelect: this.onSelect,
+      enableTabNavigation: false,
+      className: componentClassName,
+      inputValue,
+      index: '0',
+    }
+  }
+
+  renderEmpty = () => {
+    const { renderEmpty, noResultsLabel } = this.props
+    const { inputValue } = this.state
+
+    if (renderEmpty) {
+      return renderRenderPropComponent(renderEmpty)
+    }
+
+    const message = inputValue
+      ? `${noResultsLabel} for "${inputValue}"`
+      : noResultsLabel
+
+    return (
+      <EmptyItemUI className="c-ComboBoxEmpty">
+        <Text shade="muted">{message}</Text>
+      </EmptyItemUI>
+    )
+  }
+
+  renderMenuStart = () => {
+    const { renderMenuStart } = this.props
+
+    return renderMenuStart ? renderRenderPropComponent(renderMenuStart) : null
+  }
+
+  renderMenuEnd = () => {
+    const { renderMenuEnd } = this.props
+
+    return renderMenuEnd ? renderRenderPropComponent(renderMenuEnd) : null
+  }
+
+  renderFooter = () => {
+    const { renderFooter } = this.props
+    if (!renderFooter) return
+
+    return (
+      <Dropdown.Block className="c-ComboBoxFooter">
+        {renderRenderPropComponent(renderFooter)}
+      </Dropdown.Block>
+    )
+  }
```

Comment:
@ItsJonQ Tiny nitpick, but given what we now know about how these static prop arrow function are being transpiled, such that they are defined in the constructor and so each instance cannot share an underlying implementation on the prototype, should we replace all of these (except where they are serving the purpose of binding) with instance methods?

Reply:
@knicklabs You got it!!! :D